### ISSUE TITLE
an "official" recommended way to mix associated data with a derived key

### ISSUE THREAD

**Issue by oconnor663** [BLAKE3-team/BLAKE3#42]
The spec and docs for `derive_key` are very clear that the context string should be **hardcoded**, without runtime input of any kind. The fundamental purpose of `derive_key` is domain separation, and hardcoded context strings make separation more reliable: as long as your is different from mine, we're guaranteed not to interact in any way.

However, many applications do want to incorporate dynamic data into their derived keys somehow. For a silly example, maybe I have a disk encryption key, and I want to derive a separate subkey for each drive on my machine. I need to use each drive label as associated data somehow. To make this more interesting, let's say that I want a separate subkey for each partition too, so there are two distinct pieces of associated data. How should I do this?

Well, here's how I would do it:

1. Start with by using `derive_key` with a hardcoded context string, which describes what we're about to do. For example, `"Jack's disk encrypt-o-matic 2020-01-24 11:47:57 drive + partition subkeys"`. The output is a 256-bit intermediate subkey.

2. Mix in the disk label. The most natural way to do this is with `keyed_hash`. The key is the output of step 1, and the input is the disk label. The output is another 256-bit intermediate subkey.

3. Mix in the partition label. Again we'll use `keyed_hash`, with the key being the output of step 2. The output here is our final derived subkey, presumably also 256 bits.

Now, what I'd like to do is generalize this sort of thing. Some observations:

I think using `keyed_hash` is the most elegant way to do steps 2 and 3, but I want to note that it doesn't super duper matter what we do. We could concatenate the key and the associated data and just use `hash`. Because of the domain separation provided in step 1, we don't have to worry about whether our approach is going to interact badly with some other approach that someone else takes. Our keys will never interact.

Concatenating things is a little tricky, because you have to worry about ambiguities in the concatenation boundary. This doesn't really matter if one of the things you're concatenating is a fixed length, but it's something to be aware of. This is one of the reasons I prefer `keyed_hash`: there's no possible ambiguity between what's the key and what's the input.

One real world hazard that could come up here is if the application considers some associated data "optional". For example, maybe some disk has no partitions, so we don't have a partition label to use in step 3. What we _don't_ want to do is skip steps entirely, because that could introduce another type of ambiguity: which step did we skip? So there should be some kind of ruling about optional AD parameters. Maybe they can be empty, but not missing? We could also try to use a different context string at every step, but that seems awfully heavyweight.

So anyway, I'm thinking about generalizing the scheme above into some kind of broad advice for generating subkeys with associated data. I'll use this issue for keeping track of my thoughts. Hopefully other folks can chime in with theirs.

**Comment by oconnor663** [comment]
Another real world hazard is that you can't just change the associated data parameters over time. If you want to add a third parameter (let's say the filesystem type), you need to change the context string in step 1. Otherwise the key you were using for [application stuff] before is now an intermediate sublet, which is no bueno.

If we wanted a scheme that was robust to this, we could imagine some sort of "finalization" step, that distinguishes between intermediate subkeys and output subkeys. But is that really necessary? I feel like encouraging people to reuse the same context string would be a bad precedent.

**Comment by Luro02** [comment]
> What we don't want to do is skip steps entirely

You could use the rust type system for this, for example

```rust
pub struct Stage2 {
}

impl Stage2 {
    pub(crate) fn new() -> Self {
        Self {}
    }

    pub fn start_hash(data: &[u8]) {}
}

pub fn init_key(_: &str) -> Stage2 {
    // do something necessary
    Stage2::new()
}
```

This would make it impossible to call `start_hash` without calling `init_key`, so there is no possible way to skip this step.

Alternatively you could also wrap the data in a new type

```rust
struct Key(Vec<u8>);
```

which could then be passed to a function as an argument.

**Comment by oconnor663** [comment]
Agreed, providing a secure Rust API for key derivation with associated data wouldn't be too hard. But I'm thinking more about providing advice for protocol designers. Or maybe for coders using some other implementation of BLAKE3, in some other language. (Maybe even folks who aren't using BLAKE3.) I'm hoping to define some sort of Right Way to Do It in the abstract.
